### PR TITLE
Hotfix/method

### PR DESCRIPTION
**Pull request by wandmagic**
# Committer Notes

hotfix method-id bug
seen in https://github.com/usnistgov/OSCAL/issues/2059

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/usnistgov/OSCAL/blob/main/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/OSCAL/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

(For reviewers: [The wiki has guidance](https://github.com/usnistgov/OSCAL/wiki/Issue-Review) on code review and overall issue review for completeness.)

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes you made? Changes to the OSCAL website can be made in the docs/content directory of your branch.

### CODE REVIEW

File: src/metaschema/shared-constraints/allowed-values-control-group-property-name.ent [hunk 1]
```diff
@@ -1,3 +1,4 @@
 <enum xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0" value="label">A human-readable label for the parent context, which may be rendered in place of the actual identifier for some use cases.</enum>
 <enum xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0" value="sort-id">An alternative identifier, whose value is easily sortable among other such values in the document.</enum>
 <enum xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0" value="alt-identifier">An alternate or aliased identifier for the parent context.</enum>
+<enum xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0" value="method">its an assement method.</enum>
```
Comment (aj-stein-gsa) [inline src/metaschema/shared-constraints/allowed-values-control-group-property-name.ent:4]: I know I recommended to Paul because this is not ideal, but I am not sure where we would want this given discussion in the issue reported and how this prop is not constrained precisely in the assessment model. This property definition is in control common.

https://github.com/usnistgov/OSCAL/blob/v1.1.2/src/metaschema/oscal_control-common_metaschema.xml#L78-L81

I do not know where a more prudent place to put the constraint is, I think that is a part of the discussion between you and Brian, but if we do not do it here we have to duplicate a few times.

Comment (david-waltermire) [reply src/metaschema/shared-constraints/allowed-values-control-group-property-name.ent:4]: I believe the side effect here is it makes this allowed on groups which is not correct.

Comment (david-waltermire) [reply src/metaschema/shared-constraints/allowed-values-control-group-property-name.ent:4]: It's a shame these are defined in external constraints, because targeting them in multiple places would be easier, without the need to use this entity hack.

**Comment by wandmagic** [comment]
@iMichaela there were two invalid metapaths in develop i fixed them.
@brian-ruf I've placed the "method" allowed value on the part assembly. kindly let me know if this solves our issue in an acceptable manner.
@david-waltermire  I'm with you as far as external constraints, but lets save that for a bigger PR and start with a strong ADR to support the effort

File: src/metaschema/shared-constraints/allowed-values-control-group-property-name.ent (hunk 1 above)
Comment (wandmagic) [reply src/metaschema/shared-constraints/allowed-values-control-group-property-name.ent]: I've updated this to exist on just the part assembly, it should be more "surgical now"

**Comment by aj-stein-gsa** [comment]
> If we do not find a simple elegant way, the simplest solution is to remove the cardinality constraints since the world outside USG is not using them anyways:
> 
> and FedRAMP can enforce them for FedRAMP data.
> 
>  Hence NIST's conclusion (@wendellpiez and I discussed it at length beyond this issue) - better remove them from core OSCAL to allow FedRAMP and others that might want them to enforce them in any way they want .

So, just to be clear: FedRAMP would only want this constraint observed because it is the intentional structure that matches 800-53A assessment methods and objectives to align with a specific structure of how NIST maintains 800-53B controls for the given inline assessment guidance in the latter. That is really for all SP 800-53A that is provided by NIST, not FedRAMP. We choose to follow it.

So you want FedRAMP staff to propose deleting it in this pull request or another one? How will NIST ensure the assessment methods and objectives aline with controls without it during the maintenance of the 800-53A and 800-53B data in the catalogs? I only ask because about the mechanics here because automated checking of that will go away and the content in question is really maintained by NIST, not FedRAMP.

I hope that clarifies the concern. Re the proposal in https://github.com/usnistgov/OSCAL/discussions/2072, the complexity of this issue is one reason this issue and others were not included in my proposal.

**Comment by wandmagic** [comment]
I think an ADR regarding scope of this repository in regards to namespacing standards would be good.
as an OSCAL user, i expect all data in this repo to be under the default oscal namespace. Now that metaschema supports external constraints, all other namespaces should exist as external constraint modules which can be included as needed in validation. If we DO want to include other namespaces by default inside this repo, they should be explicitly documented and reason for their existence explained.

**Comment by wandmagic** [comment]
I've tested the latest commit against which is now considered valid, it allowed both method and method-id, since method appears to be deprecated, but not yet forbidden, we could add a warning constraint to illuminate this information

**Comment by iMichaela** [comment]
> > If we do not find a simple elegant way, the simplest solution is to remove the cardinality constraints since the world outside USG is not using them anyways:
> > and FedRAMP can enforce them for FedRAMP data.
> > Hence NIST's conclusion (@wendellpiez and I discussed it at length beyond this issue) - better remove them from core OSCAL to allow FedRAMP and others that might want them to enforce them in any way they want .
> 
> So, just to be clear: FedRAMP would only want this constraint observed because it is the intentional structure that matches 800-53A assessment methods and objectives to align with a specific structure of how NIST maintains 800-53B controls for the given inline assessment guidance in the latter. That is really for all SP 800-53A that is provided by NIST, not FedRAMP. We choose to follow it.
> 

@aj-stein-gsa - If we want OSCAL to succeed, we (collectively) need to ensure the entire OSCAL community is happy, not only 800-53 users ... 

> So you want FedRAMP staff to propose deleting it in this pull request or another one? How will NIST ensure the assessment methods and objectives aline with controls without it during the maintenance of the 800-53A and 800-53B data in the catalogs? I only ask because about the mechanics here because automated checking of that will go away and the content in question is really maintained by NIST, not FedRAMP.
> 

This bug has been raised in OSCAL Lobby a year ago. I believe we even had a discussion there. Others raised it too in oscal-content repo. I moved it here more recently to include it in what I was preparing for a patch release BEFORE working on prototypes to address changes and release them.

This constraints clashing  bug manifests ONLY in `local-definition` in the assessment layer.  Or at least ONLY there was observed and raised. I did not rush to fix it because I dd not want to affect FedRAMP more than any other OSCAL user, because FedRAM is ahead of others.  All I am suggesting is to be wise when fixing it. IF 800-53 will need anything specific, we can put it under the RMF `ns`, not OSCAL `ns`. We have many RMF specific constraints. 

> I hope that clarifies the concern. Re the proposal in #2072, the complexity of this issue is one reason this issue and others were not included in my proposal.

@aj-stein-gsa - If we want OSCAL to succeed, we (collectivly) need to ensure the entire OSCAL community is happy, not only 800-53 users ...

**Comment by iMichaela** [comment]
> I think an ADR regarding scope of this repository in regards to namespacing standards would be good. as an OSCAL user, i expect all data in this repo to be under the default oscal namespace. Now that metaschema supports external constraints, all other namespaces should exist as external constraint modules which can be included as needed in validation. If we DO want to include other namespaces by default inside this repo, they should be explicitly documented and reason for their existence explained.

Indeed.While a cleanup will be disruptive, it is very much needed and planned.  **We have 6 (six) RMF-specific and 14 ( fourteen! FedRAMP-specific constraints**. These should not be part of core OSCAL because an international standards body **will not accept** them. As we prepare OSCAL for international standardization in the year (or two) to come, and as other NIST data sets will be formatted  in OSCAL and supported, this approach of including `ns` specific constraints into core OSCAL does not scale, hence the OSCAL community (the foundation) intention to establish and maintain a registry .

**Comment by wandmagic** [comment]
this is how I'm testing the changes
oscal-cli metaschema validate-content src/test/content/example.xml -m ./src/metaschema/oscal_assessment-plan_metaschema.xml

re-introduced cardinality constraints with allowing at least one of method-id or method

**Comment by wendellpiez** [comment]
Just a couple of random observations:

- As @david-waltermire remarks, this would all be better done not in the standard modules but in layered modules. I wonder if there is agreement on removing most constraints from the core models, on this principle.
- Incremental improvements in the meantime (to the core modules) are both necessary and appropriate.
- Such improvements could surely include refactoring (the `let` feature would support this nicely) that would permit removing those entity references, which are a thorn in the side of developers.

To summarize: :+1: to factoring constraints out of the core, at least where possible.

**Comment by wendellpiez** [comment]
Continuing to read the comments on this Issue, I see some good points made by @aj-stein-gsa among others.

Would it be helpful to conceive of a constraints layer that would specifically address SP 800-53/A? (This could be taken quite a long way as [Schematrons for this data](https://github.com/usnistgov/oscal-content/tree/main/src/nist.gov/SP800-53/rev5/xml) already show.)

Separately, would it be helpful to conceive of a constraints set that would specifically address profiles and catalogs with respect to OSCAL profile resolution semantics? This might have to be a long-term goal since it requires alignment, with both the profile resolution specification and unit/conformance testing for profile resolvers (because it is about error handling).

But even if it were never completed, it seems to me such a rules set could be useful to start on, to the extent that such rules become salient. Seeing the applicable rules together is itself instructive.

The bigger idea is to scope the definition and maintenance of the rules to the document sets / governance / requirements being addressed. So SP800-53-based RMF in OSCAL has a set of rules making everything predictable.

FWIW, as other catalogs supporting RMF come online, this question could become more complicated.

**Comment by iMichaela** [comment]
> > @aj-stein-gsa - If we want OSCAL to succeed, we (collectivly) need to ensure the entire OSCAL community is happy, not only 800-53 users ...
> 
> I do not know where I said OSCAL _only_ has to be for use cases with SP 800-53 control requirements, I said that NIST maintains official copies of SP 800-53A and SP 800-53B in the officially published OSCAL catalog. Removing these constraints, not correcting them, means detecting a misalignment or incorrect prop definition to have the control data and assessment data "match each other" (for lack of a better term) will no longer be easily done. It makes NIST's work on their own catalog require effort for something that is not automatically checked, so I am calling that out.
> 
> If you want them removed, that is fine, I am just underlining the result of that choice to make sure you and your team are fine with that.

I did not say you said it.. I stated it as a general direction, even in the context of NIST being the 800-53 maintainer. And justified it by the CSF v2 upcoming release (which would have been out if we would not have shifted the focus to OSCAL 1.1.3 to support FEdRAMP's deadline.

**Comment by iMichaela** [comment]
The conversation here deviated towards the strategic approach and it is hard to filter the PR-specific messages. Can we please move that conversation to the Discussion board and try tasty focus here on fixing the bug for now in the safest way possible so we can release 1.1.3?

**Comment by wandmagic** [comment]
> The conversation here deviated towards the strategic approach and it is hard to filter the PR-specific messages. Can we please move that conversation to the Discussion board and try tasty focus here on fixing the bug for now in the safest way possible so we can release 1.1.3?

so should we remove the cardinality constraints, or allow cardinality of either method or method id or only method id?

**Comment by aj-stein-gsa** [comment]
OK to focus on the PR: 

We will try to adjust accordingly, we understand everyone has priorities. But back to the important question: do you recommend deleting the constraint in an updated PR now? Do you need time to think about it?

Details below, but everything below this statement is context and not urgent. Per your request, I will move release convo to 1.1.3 discussion in #2072.

> I did not say you said it.. I stated it as a general direction, even in the context of NIST being the 800-53 maintainer.

I am asking what you want for the context of the review. You proposed that one simple solution is just to delete them in https://github.com/usnistgov/OSCAL/pull/2076#issuecomment-2483031472. I am not disagreeing, it is simple and will make things convenient for me (I said me, not FedRAMP). Long-term, I am not sure it will make the work of maintaining the 800-53 catalog and related examples of ideal use with SP 800-53A and SP 800-53B _easier for the NIST OSCAL Team_. I am trying to flag that.

That being said, I still asked: is this what you and the OSCAL Team want, to adjust the PR? I asked and so did Paul in https://github.com/usnistgov/OSCAL/pull/2076#issuecomment-2483576857. I am providing context for the question asked.

> And justified it by the CSF v2 upcoming release (which would have been out if we would not have shifted the focus to OSCAL 1.1.3 to support FEdRAMP's deadline.

OK, I will reiterate what I expressly asked, which is public record, in https://github.com/usnistgov/OSCAL/discussions/2072#discussioncomment-11267710. I did not say I or FedRAMP want or need a deadline of 18 November 2024. I asked a series of questions about whether the proposal was sound, could the team commit to a _tentative_ release date, or if they could not at all. I know I cannot set the deadline for the team, I was asking if an estimate was possible. Had you had mentioned the CSF 2.0 release as a priority and a decision for no deadline or a later one, we would accept that. I was asking for your perspective, not requiring one of the options and a specific deadline. I am trying to be incredibly transparent and fair in these inquiries.

**Comment by wandmagic** [comment]
this PR is getting stale, closing

File: src/metaschema/oscal_control-common_metaschema.xml [hunk 2]
```diff
@@ -77,7 +77,9 @@
                 </model>
                 <constraint>
                         <allowed-values target="prop[has-oscal-namespace('http://csrc.nist.gov/ns/oscal')]/@name">
-            &allowed-values-control-group-property-name;
+                                &allowed-values-control-group-property-name;
+                                <enum xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0" value="method-id">its an assement method.</enum>
```
Comment (iMichaela) [inline src/metaschema/oscal_control-common_metaschema.xml:81]: This constraint has poor description AND adds an allowed value "method-id" which is proposed to be removed from the  assessment layer.

File: src/metaschema/oscal_assessment-common_metaschema.xml [hunk 3]
```diff
@@ -71,7 +71,7 @@
...
          <has-cardinality target="part[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name=('assessment','assessment-method')]/prop[has-oscal-namespace(('http://csrc.nist.gov/ns/oscal','http://csrc.nist.gov/ns/rmf')) and @name='method']" min-occurs="1" max-occurs="1" />
          <has-cardinality target="part[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name=('assessment','assessment-method')]/part[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name=('objects','assessment-objects')]" min-occurs="1" max-occurs="1" />
-         <has-cardinality target="part[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name=('objective','assessment-objective')]/prop[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name='method-id']" min-occurs="1" />
+         <has-cardinality id='objective-has-method' target="part[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name=('objective','assessment-objective')]/prop[has-oscal-namespace('http://csrc.nist.gov/ns/oscal') and @name=('method','method-id')]" min-occurs="1" />
```
Comment (iMichaela) [inline src/metaschema/oscal_assessment-common_metaschema.xml:74]: Why or how would this change fix the conflicting by adding an option for name to be either "method" or "method-id".  I would like to see examples and explanation of that is aimed here to accomplish.
